checks/ban-mutable-static/src/main/java/edu/ucr/cs/riple/banmutablestatic/BanMutableStatic.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Nima Karimipour
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package edu.ucr.cs.riple.banmutablestatic;
+
+import static com.google.errorprone.BugPattern.SeverityLevel.SUGGESTION;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol;
+import javax.lang.model.element.Modifier;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+    name = "BanMutableStatic",
+    altNames = {"BansMutableStaticFields"},
+    summary = "Bans mutable static field declaration",
+    tags = BugPattern.StandardTags.STYLE,
+    severity = SUGGESTION)
+@SuppressWarnings("BugPatternNaming")
+public class BanMutableStatic extends BugChecker implements BugChecker.VariableTreeMatcher {
+
+  public final String ERROR_MESSAGE =
+      "Variable %s is declared as a mutable static field, should be replaced with: final static %s %s";
+  private final String LINK_URL = "https://github.com/nimakarimipour/NullAwayAnnotator";
+
+  public BanMutableStatic() {}
+
+  @Override
+  public Description matchVariable(VariableTree tree, VisitorState state) {
+    boolean isStatic = tree.getModifiers().getFlags().contains(Modifier.STATIC);
+    boolean isMutable = !tree.getModifiers().getFlags().contains(Modifier.FINAL);
```

Comment:
This only covers "shallow" mutability, but "deep" mutability is also important.  E.g., if you have a `static final HashMap` that is mutable even though it is final, since you can mutate the contents of the `HashMap`.  So this check will catch some of the bad cases, but not all.

Reply:
Thanks for the note. Is this something we would like to avoid ? Is this also problematic If multiple instances are running and is guaranteed that the list is singleton and shared among them ?

Reply:
Absolutely we need to avoid it.  It causes huge problems like data races and memory leaks.  I don't understand the singleton list question; if you know ahead of time it's a singleton why do you need a list?

Reply:
Sure, will also cover them. Thank you. By singleton I meant only __one__ instance of list which can have arbitrarily number of elements. I will cover cases you mentioned as well.

Reply:
If the contents of the list change over time then it is bad 🙂 If it's an `ImmutableList` it's ok

Reply:
Haven't had a chance yet to do a full review, but, regarding the `BanMutableStatic` checker only... why not make it so it requires that static fields are both `final` _and_ annotated `@Immutable`, then leverage this existing EP checker: https://errorprone.info/bugpattern/Immutable

Reply:
@lazaroclapp I think the`@Immutable` annotation applies to classes, not fields?  If we really wanted to catch this we could check that the type of the field is `@Immutable` maybe by re-using code from https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/threadsafety/ImmutableAnalysis.java

Reply:
@lazaroclapp @msridhar I looked at this code, all I can access is `checkForImmutability` method which can be called on class tree. Would you please tell me what you mean by reusing this code. I think we have two options, either override match class method and call it from there, or copy codes from `areFieldsImmutable` and keep overriding `matchVariable` method. Thank you.

Reply:
At this point I think this is low priority, and I wouldn't spend time improving the mutable static checker.  Given the size of the scanner, it shouldn't be hard to just search for all static fields using IntelliJ and check them manually.

Reply:
@msridhar Ok sure sounds good. I will keep the checker as is. I may work on it in future if there was nothing outstanding to do since it seems like an interesting side project :).